services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt

```diff
@@ -113,38 +167,287 @@ class HakemusServiceITest : DatabaseTest() {
+    open inner class UpdateHakemus {
+
+        private val intersectingArea =
+            ApplicationFactory.createApplicationArea(
+                name = "area",
+                geometry =
+                    "/fi/hel/haitaton/hanke/geometria/intersecting-polygon.json".asJsonResource()
+            )
+
+        private val notInHankeArea =
+            ApplicationFactory.createApplicationArea(
+                name = "area",
+                geometry = GeometriaFactory.polygon
+            )
+
+        @Test
+        fun `throws exception when the application does not exist`() {
+            assertThat(applicationRepository.findAll()).isEmpty()
+
+            assertThrows<ApplicationNotFoundException> {
+                hakemusService.updateHakemus(
+                    1234,
+                    HakemusUpdateRequestFactory.createFilledJohtoselvityshakemusUpdateRequest(),
+                    USERID
+                )
+            }
+        }
+
+        @Test
+        fun `throws exception when the application has been sent to Allu`() {
+            val application = createHakemus { alluid = 21 }
 
-    private fun Assert<CustomerWithContactsResponse>.isCompanyCustomerWithOneContact(
-        orderer: Boolean
-    ) {
-        prop(CustomerWithContactsResponse::customer)
-            .prop(CustomerResponse::type)
-            .isEqualTo(CustomerType.COMPANY)
-
-        prop(CustomerWithContactsResponse::contacts)
-            .single()
-            .prop(ContactResponse::orderer)
-            .isEqualTo(orderer)
+            val exception =
+                assertThrows<ApplicationAlreadySentException> {
+                    hakemusService.updateHakemus(
+                        application.id!!,
+                        HakemusUpdateRequestFactory.createFilledJohtoselvityshakemusUpdateRequest(),
+                        USERID
+                    )
+                }
+
+            assertThat(exception.message)
+                .isEqualTo("Application is already sent to Allu, id=${application.id}, alluid=21")
+        }
+
+        @Test
+        fun `does not create a new audit log entry when the application has not changed`() {
+            val application = createHakemus()
+            val originalAuditLogSize = auditLogRepository.findByType(ObjectType.APPLICATION).size
+
+            hakemusService.updateHakemus(
+                application.id!!,
+                HakemusUpdateRequestFactory
+                    .createJohtoselvityshakemusUpdateRequestFromApplicationEntity(application),
+                USERID
+            )
+
+            val applicationLogs = auditLogRepository.findByType(ObjectType.APPLICATION)
+            assertThat(applicationLogs).hasSize(originalAuditLogSize)
+        }
+
+        @Test
+        fun `throws exception when there are invalid geometry in areas`() {
+            val application = createHakemus()
+
+            val exception =
+                assertThrows<ApplicationGeometryException> {
```

Comment:
I think we should stick with one assertion library as much as possible.
```suggestion
                assertFailure {
```

The class of the exception can be checked with `exception.hasClass()`.